### PR TITLE
Fix-5358: XACML policy based scope validation when token and code issuing phase

### PR DESCRIPTION
**Pull request by Buddhimah**
Merge this PR after merging the related PR

### CODE REVIEW

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java [hunk 1]
```diff
@@ -64,80 +68,84 @@ public boolean validateScope(AccessTokenDO accessTokenDO, String resource) throw
...
-        String authzUser = accessTokenDO.getAuthzUser().getUserName();
-        boolean isValidated = false;
-        try {
-            FrameworkUtils.startTenantFlow(accessTokenDO.getAuthzUser().getTenantDomain());
-            String consumerKey = accessTokenDO.getConsumerKey();
-            OAuthAppDO authApp = OAuth2Util.getAppInformationByClientId(consumerKey);
+        String consumerKey = accessTokenDO.getConsumerKey();
+        return validateScope(accessTokenDO.getScope(), accessTokenDO.getAuthzUser(), consumerKey,
+                XACMLScopeValidatorConstants.ACTION_VALIDATE, resource);
+    }
```
Comment (Abilashini) [inline component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: Add a blank line at the end of a method

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java [hunk 2]
```diff
@@ -64,80 +68,84 @@ public boolean validateScope(AccessTokenDO accessTokenDO, String resource) throw
...
+        FrameworkUtils.startTenantFlow(authenticatedUser.getTenantDomain());
+        if (StringUtils.isNotEmpty(consumerKey)) {
+            try {
+
+                OAuthAppDO oAuthAppDO = getOAuthAppDO(consumerKey);
+                String request = createRequest(scopes, authenticatedUser, oAuthAppDO, action, resource);
+                isValid = isRequestPermit(request, oAuthAppDO, authenticatedUser.toFullQualifiedUsername());
+
+            } catch (InvalidOAuthClientException e) {
+                throw new IdentityOAuth2Exception(String.format("Exception occurred when getting app information for " +
```
Comment (Abilashini) [inline component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: The same message repeated 2 times

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java [hunk 3]
```diff
@@ -64,80 +68,84 @@ public boolean validateScope(AccessTokenDO accessTokenDO, String resource) throw
...
     }
 
     /**
-     * Creates RequestDTO object for XACML request with the parameters retrieved from the access token
-     *
-     * @param accessTokenDO access token
-     * @param authApp       OAuth app
-     * @param resource      resource
-     * @return RequestDTO
+     *  Creates XACML Request string with the parameters retrieved from the request.
```
Comment (Abilashini) [inline component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: Alignment issue

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java [hunk 4]
```diff
@@ -149,14 +157,71 @@ private RequestDTO createRequestDTO(AccessTokenDO accessTokenDO, OAuthAppDO auth
...
+     * @param oAuthAppDO Application.
+     * @param authzUser Fully qualified name of the user.
+     * @return Returns true if the XACML response is permit or NotApplicable. Else returns false.
+     * @throws IdentityOAuth2Exception Exception
+     */
+    private boolean isRequestPermit(String request, OAuthAppDO oAuthAppDO, String authzUser) throws IdentityOAuth2Exception {
+
+        boolean permit = false;
+        try {
+            String responseString = OAuthScopeValidatorDataHolder.getInstance().getEntitlementService().getDecision
```
Comment (Abilashini) [inline component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: If the line exceeds 120 characters, then move '.getDecision(request);' together to the next line

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java [hunk 5]
```diff
@@ -149,14 +157,71 @@ private RequestDTO createRequestDTO(AccessTokenDO accessTokenDO, OAuthAppDO auth
...
+                    (request);
+            if (log.isDebugEnabled()) {
+                log.debug("XACML scope validation response :\n" + responseString);
+            }
+            String response = extractDecisionFromXACMLResponse(responseString);
+            if (isResponseNotApplicable(response)) {
+                log.warn(String.format(
+                        "No applicable rule for service provider '%s@%s'. Add an validating policy (or unset Scope " +
```
Comment (Abilashini) [inline component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: Can be merged with the previous line. split to the next line at the end of 120 characters and merge with the below line

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java [hunk 6]
```diff
@@ -149,14 +157,71 @@ private RequestDTO createRequestDTO(AccessTokenDO accessTokenDO, OAuthAppDO auth
...
+    /**
+     * Validates the XACML request using XACML engine with the parameters authApp and authzUser, and returns whether
+     * to permit or not.
+     * @param request XACML request.
+     * @param oAuthAppDO Application.
+     * @param authzUser Fully qualified name of the user.
+     * @return Returns true if the XACML response is permit or NotApplicable. Else returns false.
+     * @throws IdentityOAuth2Exception Exception
+     */
+    private boolean isRequestPermit(String request, OAuthAppDO oAuthAppDO, String authzUser) throws IdentityOAuth2Exception {
```
Comment (Abilashini) [inline component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: Line exceeds 120 characters

Comment (Buddhimah) [reply component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: done

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java (hunk 5 above)
Comment (Buddhimah) [reply component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: done

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java (hunk 4 above)
Comment (Buddhimah) [reply component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: done

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java (hunk 3 above)
Comment (Buddhimah) [reply component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: done

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java (hunk 2 above)
Comment (Buddhimah) [reply component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: done

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java (hunk 1 above)
Comment (Buddhimah) [reply component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: done

File: component/client-handler/src/test/resources/dbscripts/identity.sql [hunk 7]
```diff
@@ -7,21 +7,22 @@ INSERT INTO IDN_BASE_TABLE values ('WSO2 Identity Server');
...
+            OAUTH_VERSION VARCHAR(128),
+            CALLBACK_URL VARCHAR(1024),
             GRANT_TYPES VARCHAR (1024),
             PKCE_MANDATORY CHAR(1) DEFAULT '0',
             PKCE_SUPPORT_PLAIN CHAR(1) DEFAULT '0',
             APP_STATE VARCHAR (25) DEFAULT 'ACTIVE',
-            USER_ACCESS_TOKEN_EXPIRE_TIME BIGINT DEFAULT 3600000,
-            APP_ACCESS_TOKEN_EXPIRE_TIME BIGINT DEFAULT 3600000,
-            REFRESH_TOKEN_EXPIRE_TIME BIGINT DEFAULT 84600000,
+            USER_ACCESS_TOKEN_EXPIRE_TIME BIGINT DEFAULT 3600,
```
Comment (mefarazath) [inline component/client-handler/src/test/resources/dbscripts/identity.sql]: We are changing the default value from 3600000 -> 3600. Is this correct?

Comment (Buddhimah) [reply component/client-handler/src/test/resources/dbscripts/identity.sql]: This is the same table structure that was used in IS-5.8.0

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java [hunk 8]
```diff
@@ -189,20 +190,20 @@ private String createRequest(String[] scopes, AuthenticatedUser authenticatedUse
...
+                    .getDecision(request);
             if (log.isDebugEnabled()) {
                 log.debug("XACML scope validation response :\n" + responseString);
             }
             String response = extractDecisionFromXACMLResponse(responseString);
             if (isResponseNotApplicable(response)) {
-                log.warn(String.format(
-                        "No applicable rule for service provider '%s@%s'. Add an validating policy (or unset Scope " +
-                                "Validation using XACMLScopeValidator) to fix this warning.",
+                log.warn(String.format("No applicable rule for service provider '%s@%s'. Add an validating policy "
```
Comment (Buddhimah) [inline component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: add an -> add a

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java [hunk 9]
```diff
@@ -115,8 +116,7 @@ private boolean validateScope(String[] scopes, AuthenticatedUser authenticatedUs
                 isValid = isRequestPermit(request, oAuthAppDO, authenticatedUser.toFullQualifiedUsername());
 
             } catch (InvalidOAuthClientException e) {
-                throw new IdentityOAuth2Exception(String.format("Exception occurred when getting app information for " +
-                        "client id %s of user %s. Error occurred when retrieving corresponding app for this specific" +
+                throw new IdentityOAuth2Exception(String.format("Error occurred when retrieving corresponding app for this specific" +
```
Comment (piraveena) [inline component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: Here you need to pass variables for client and user using %s

**Comment by mefarazath** [comment]
Travis build has failed. Does this depend on a newly introduced API?

**Comment by Buddhimah** [comment]
this depends on auth-oauth newly introduced API. did a version bump

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/test/java/org/wso2/carbon/identity/oauth2/validators/XACMLScopeValidatorTest.java [hunk 10]
```diff
@@ -91,42 +97,93 @@ public IObjectFactory getObjectFactory() {
...
+
     }
 
-    @Test
-    public void testCreateRequestDTO() throws Exception {
+    @DataProvider(name = "createRequestObj")
+    public Object[][] createRequestObj() {
+
+        return new Object[][]{
+                // Create XACML request string for token validation phase
```
Comment (JKAUSHALYA) [inline component/org.wso2.carbon.identity.oauth2.validators.xacml/src/test/java/org/wso2/carbon/identity/oauth2/validators/XACMLScopeValidatorTest.java]: All comments should end with a full stop.

**Comment by Buddhimah** [comment]
Fixed Comments : https://github.com/wso2-extensions/identity-oauth-addons/pull/49

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/test/java/org/wso2/carbon/identity/oauth2/validators/XACMLScopeValidatorTest.java [hunk 11]
```diff
@@ -91,42 +97,93 @@ public IObjectFactory getObjectFactory() {
...
         authenticatedUser.setUserName(ADMIN_USER);
         accessTokenDO = new AccessTokenDO();
         accessTokenDO.setConsumerKey("consumer-key");
         accessTokenDO.setAuthzUser(authenticatedUser);
         accessTokenDO.setScope(scopeArray);
         authApp = new OAuthAppDO();
         authApp.setApplicationName(APP_NAME);
+
+        oauth2AccessTokenReqDTO = new OAuth2AccessTokenReqDTO();
+        oauth2AccessTokenReqDTO.setClientId("consumer-key");
```
Comment (sachiniWettasinghe) [inline component/org.wso2.carbon.identity.oauth2.validators.xacml/src/test/java/org/wso2/carbon/identity/oauth2/validators/XACMLScopeValidatorTest.java]: better to add a constant for "consumer-key"

File: component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java [hunk 12]
```diff
@@ -64,80 +68,85 @@ public boolean validateScope(AccessTokenDO accessTokenDO, String resource) throw
...
+     * @throws IdentityOAuth2Exception by an Underline method.
+     */
+    private boolean validateScope(String[] scopes, AuthenticatedUser authenticatedUser, String consumerKey,
+                                  String action, String resource) throws IdentityOAuth2Exception {
+
+        boolean isValid = false;
+        FrameworkUtils.startTenantFlow(authenticatedUser.getTenantDomain());
+        if (StringUtils.isNotEmpty(consumerKey)) {
+            try {
+
```
Comment (sachiniWettasinghe) [inline component/org.wso2.carbon.identity.oauth2.validators.xacml/src/main/java/org/wso2/carbon/identity/oauth2/validators/xacml/XACMLScopeValidator.java]: remove redundant lines

**Pull request merged**